apps/dash-daq-tektronix350/app.py

```diff
@@ -0,0 +1,542 @@
+def update_frequency_display(value):
+    fgen.set_frequency(value)
+    return value
+
+
+@app.callback(Output("frequency-display", "value"), [Input("amplitude-input", "value")])
+def update_amplitude_display(value):
+    fgen.set_amplitude(value)
+    return value
+
+
+@app.callback(Output("offset-display", "value"), [Input("offset-input", "value")])
+def update_offset_display(value):
+    fgen.set_offset(value)
+    return value
+
+
+@app.callback(Output("offset-display", "value"), [Input("function_type", "value")])
+def update_fgen_wave(value):
+    fgen.set_wave(value)
+    return value
+
+
+# Callbacks graph and graph info
+@app.callback(
+    Output("graph_info", "children"),
+    [Input("oscope-graph", "figure"), Input("tabs", "value")],
+)
+def update_info(_, value):
+    if "" + str(value) in runs:
+        return runs["" + str(value)][1]
+    return "-"
+
+
+@app.callback(
+    Output("oscope-graph", "figure"),
+    [Input("update-oscope", "n_intervals"), Input("tabs", "value")],
+)
+def update_output(_, value):
+    global tab
+    time = np.linspace(-0.000045, 0.000045, 1e3)
+    zero = dict(
+        data=[dict(x=time, y=[0] * len(time), marker={"color": "#2a3f5f"})],
+        layout=go.Layout(
+            xaxis={
+                "title": "s",
+                "color": "#506784",
+                "titlefont": dict(family="Dosis", size=15),
+            },
+            yaxis={
+                "title": "Voltage (mV)",
+                "color": "#506784",
+                "titlefont": dict(family="Dosis", size=15),
+            },
+            margin={"l": 40, "b": 40, "t": 0, "r": 50},
+            plot_bgcolor="#F3F6FA",
+        ),
+    )
+
+    if tab is not value:
+        if "" + str(value) in runs:
+            tab = value
+            return runs["" + str(value)][0]
+        tab = value
+        return zero
+
+    else:
+        figure = {
+            "data": osc.get_data(),
+            "layout": go.Layout(
+                xaxis={
+                    "title": "s",
+                    "color": "#506784",
+                    "titlefont": dict(family="Dosis", size=15),
+                },
+                yaxis={
+                    "title": "Voltage (mV)",
+                    "color": "#506784",
+                    "titlefont": dict(family="Dosis", size=15),
+                    "autorange": False,
+                    "range": [-10, 10],
+                },
+                margin={"l": 40, "b": 40, "t": 0, "r": 50},
+                plot_bgcolor="#F3F6FA",
+            ),
+        }
+
+        runs["" + str(value)] = (
+            figure,
+            str(fgen.get_wave())
+            + " | "
+            + str(fgen.get_frequency())
+            + "Hz"
+            + " | "
+            + str(fgen.get_amplitude())
+            + "mV"
+            + " | "
+            + str(fgen.get_offset())
+            + "mV",
+        )
+
+        # wait to update the runs variable
+        sleep(0.10)
```

Comment:
Is the sleep function necessary for this app?

Reply:
@christopherjeon Yep! We wanted to give the instrument some time to communicate.